src/NuGet.Clients/NuGet.CommandLine/Commands/SearchCommand.cs

```diff
@@ -83,6 +84,25 @@ public override async Task ExecuteCommandAsync()
                 PackageSearchResource resource = await repository.GetResourceAsync<PackageSearchResource>();
 
                 if (resource is null)
+                {
+                    taskList.Add((null, source));
+                    continue;
+                }
+
+                taskList.Add((Task.Run(() => resource.SearchAsync(
```

Comment:
If I am reading it correctly, this will parallelize all the queries at once.
It might be worth throttling them, like https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs#L34 for example.

https://github.com/NuGet/NuGet.Client/blob/efa928b55549ce0a8d56bb82fd4d060e62b8ecfc/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs#L73-L86

Reply:
Also, overuse of `Task.Run` is an antipattern (in general, I'm not making judgment on this PR).

* It's not being done in this example, but doing something like:
```cs
for (int i = 0; i < whatever; i++)
{
  tasks.Add(Task.Run() => DoSomething(i));
}
```
This can give wrong results because of the way that value types are captured in the anonymous method.

* If the method being called inside `Task.Run` is properly async and does not do much CPU work before reaching the first async part, then `Task.Run` does nothing more than increase latency

However, I think (I'm not 100% sure) that our local feed implementation is not proper async, so running `resounce.SearchAsync` on a local feed without `Task.Run` will probably not run things in parallel.

Anyway, this comment is an FYI since you're an intern and this is supposed to be a learning opportunity. Performance, async and parallelism is hard and there are lots of gotchas, and reading existing code won't teach you these things.

Reply:
@nkolev92 Is there (or could there be) a significant performance improvement if I limit the number of active threads to the `Environment.ProcessorCount` instead? In this particular case, completing an active task when at full capacity would mean that we wait to complete the first task that was added to the list, since we want to print the results from the sources in the order they were specified. Would this affect/invalidate that performance improvement, or is it still worth it to do things that way?

Also, thanks for explaining all of that @zivkan!

Reply:
It depends. 

The impact is particularly felt on machines with low resources. 
Trying to make too many remote calls can easily lead to "Too many open files" on linux/mac failures and affect reliability. 

Now search is unlikely to be used on such machines, but you get the sentiment.

Why does the order of the sources matter? 
Wouldn't it a better experience to return results whenever they are available?

Reply:
If a user specifies the sources themselves then it might not matter as much, but if the sources are picked up from nuget.config, wouldn't we want to display the results from a source like NuGet.org before other sources? Also, let's say some sources just return a null result because they weren't able to perform the search, their tasks might complete earlier than 'more relevant' sources might, and there might be a few "No results found." messages before a user even comes across useful results. Is this something we should be taking into account, or should we look to print results as they become available regardless of these visibility-related factors?

Reply:
> if the sources are picked up from nuget.config, wouldn't we want to display the results from a source like NuGet.org before other sources? 

No, there's no source preference. Why should nuget.org matter more than other sources? 

> Also, let's say some sources just return a null result because they weren't able to perform the search, their tasks might complete earlier than 'more relevant' sources might, and there might be a few "No results found." messages before a user even comes across useful results.

You do group them by sources right? So "No results found" will only be shown for that source. 

> because they weren't able to perform the search,

What happens today if a source is unavailable? 
I think the search call might throw and terminate the whole operation.

Reply:
I think it would be a really poor user experience if a user does two consecutive searches, and gets the results for those in a different order each time.

Also, like, I totally think we should fuck up processors with tasks. They're lightweight! That's the point! I don't think we should do any throttling, and I continue to assert that it'll be more efficient without any throttling, even on single-cpu machines. :)

Reply:
> Also, like, I totally think we should fuck up processors with tasks. They're lightweight

It's not about the tasks, it's about `Task.Run`. You can starve the threadpool. I'm not saying that'll happen here, it won't. 
Leaving it as it won't change much.
The challenge is that some of our protocol code has custom built timeouts, where we can't account for too many continuations being scheduled.

Reply:
I am not 100% sure about the advantages yet, can we use [`Paralle.ForEach`](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/data-parallelism-task-parallel-library) here for `foreach (PackageSource source in listEndpoints)` loop instead of 2 `foreach` loops ?

Reply:
Server team uses this pattern:

```cs
// I have a bunch of items that I need to process in parallel:
var items = new List<object>();

// I'll create a bag of work that can be processed in parallel.
var work = new ConcurrentBag<object>(items);

// I'll setup up to 32 tasks to process the work in parallel.
// If I have fewer than 32 items to process, I'll use fewer tasks.
// For example, I'll only create 2 tasks if there are two items to process.
Enumerable
  .Repeat(0, Math.Min(items.Count, 32))
  .Select(async _ =>
  {
    // Take an item from the bag to process.
    while (work.TryTake(out var item))
    {
      // Process the item...
    }
  });
```

Reply:
> Also, like, I totally think we should fuck up processors with tasks. They're lightweight! That's the point! I don't think we should do any throttling, and I continue to assert that it'll be more efficient without any throttling, even on single-cpu machines. :)

The .NET thread pools queue tasks. Flooding the threadpool with tasks will result in the thread pool creating more threads to try and keep up with the backlog. Things like timers will get delayed. As a rule of thumb, you should avoid doing this if there's an easy work around available.

Reply:
> Also, overuse of `Task.Run` is an antipattern (in general, I'm not making judgment on this PR).
> 
> * It's not being done in this example, but doing something like:
> 
> ```cs
> for (int i = 0; i < whatever; i++)
> {
>   tasks.Add(Task.Run() => DoSomething(i));
> }
> ```
> 
> This can give wrong results because of the way that value types are captured in the anonymous method.
> 
> * If the method being called inside `Task.Run` is properly async and does not do much CPU work before reaching the first async part, then `Task.Run` does nothing more than increase latency
> 
> However, I think (I'm not 100% sure) that our local feed implementation is not proper async, so running `resounce.SearchAsync` on a local feed without `Task.Run` will probably not run things in parallel.
> 
> Anyway, this comment is an FYI since you're an intern and this is supposed to be a learning opportunity. Performance, async and parallelism is hard and there are lots of gotchas, and reading existing code won't teach you these things.

@zivkan So it's better to be used on synchronous DoSomething rather than async DoSomething?
Also is there correctness problem?

Reply:
sorry, no, I should have written `DoSomethingAsync`. If the method was synchronous, then calling it without `Task.Run` will run the entire loop sequentially, not in parallel.

As for correctness, I'm not sure what you're referring to. But anything parallel/multi-threading related is littered with many correctness potential issues. I'd need more context to comment further, but if it's not related to something specific to this PR, it's better to do elsewhere. It's hard enough to complete a PR without a whole lot of off-topic conversations going on :)